### PR TITLE
ENH allow caching coroutine functions

### PR DESCRIPTION
**Pull request by gsakkis**
Addresses #889, allowing coroutine functions (available in Python 3.5+) to be decorated with `joblib.memory`. In short, the decorated function's `__call__` and `call_and_shelve` are coroutine functions that when `await`ed return the output or `MemorizedResult`, respectively, that would normally return if the original function was not a coroutine. A small subset of tests were ported from `test_memory` to their asyncio equivalent to verify the expected behavior.

In order to keep the asyncio-specific implementation as small as possible, I had to refactor a large portion of `memory.py` first. It's a separate commit so it can be reviewed and tested independently.

### CODE REVIEW

**Comment by gsakkis** [comment]
Any feedback on this? Btw as commented in another [PR](https://github.com/joblib/joblib/pull/897#issuecomment-504623424), the AppVeyor failure looks unrelated.

**Comment by pierreglaser** [comment]
Thanks for putting this together. I can make a pass on it if @tomMoral or @ogrisel are on board with this PR.

**Comment by WestXu** [comment]
It's been about a year since last update on this merge. Is this repo abandoned somehow? What's the substitute library guys plz?

**Comment by ckchow** [comment]
joblib.memory is great but it is painful that it doesn't work quite right with async.

**Comment by davidbrochart** [comment]
I might be interested in using `joblib.memory` to automatically cache the execution of Jupyter notebook cells in [akernel](https://github.com/davidbrochart/akernel). It already supports reactive programming, but that doesn't play well with big data. Caching cell execution would allow to "restart and run all" basically for free.
`akernel` does some AST transformation on the executed code, and already wraps the cell code into an async function. It could use `joblib.memory` if coroutines were supported.
Any plan on continuing that effort?

**Comment by leshchenko1979** [comment]
You can try using `perscache` (https://github.com/leshchenko1979/perscache) - it supports persistent caching for async functions.

**Comment by rundef** [comment]
I wish this was part of the joblib's package !

**Comment by gsakkis** [comment]
I'm not using joblib for some years now but I could try rebasing this PR if there's interest in merging it, otherwise just lmk and I'll close it.
cc @ogrisel

**Comment by tomMoral** [comment]
I think this could be a nice addition, if you are up to rebasing the PR, i can take a look to help merge it.

**Comment by gsakkis** [comment]
@tomMoral i rebased it, please take a look. Not sure what's the issue with the pypy timeout, the related tests are passing.

**Comment by tomMoral** [comment]
Thanks a lot for rebasing and sorry for the delay in answering, I will take a look at this PR.

I could not look at the pypy timeout but most probable is that it is due to a random failure of the pipeline on `pypy` that is unrelated to your changes.
I relaunched the tests to see the status here.

File: joblib/test/test_memory_async.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+from joblib.test.common import np, with_numpy
+from joblib.testing import raises
+
+from .test_memory import (corrupt_single_cache_item,
+                          monkeypatch_cached_func_warn)
+
+
+async def f(x, y=1):
+    await asyncio.sleep(0.1)
+    return x ** 2 + y
```
Comment (tomMoral) [inline joblib/test/test_memory_async.py:18]: This function is only used in `test_call_and_shelve_async`, so probably better to define it in the test?

**Review by tomMoral — APPROVED** [review]
This looks quite good!

I did a change on the name `path` -> `call_id`, did not expect this change to be in so many places :sweat_smile: 
Some cosmetic changes but overall I think this PR is very nice, thanks for the refactoring which makes the code much easier to read.

**Comment by gsakkis** [comment]
@tomMoral thanks for the review! I noticed in some places you rename path to cache_id and others to call_id; is this intentional?

**Comment by tomMoral** [comment]
> @tomMoral thanks for the review! I noticed in some places you rename path to cache_id and others to call_id; is this intentional?

no, I think `call_id` was my goal 😁

**Comment by tomMoral** [comment]
The failure is due to outdated install of sklearn's nightly build, which is fixed in #1567 
I will merge this one once the CI is green.

**Comment by tomMoral** [comment]
Thanks a lot @gsakkis !! That's a nice feature!

Sorry for the (very) slow process and thanks for keeping up with it :pray:

**Pull request merged**